### PR TITLE
Translation of llvm.dbg.declare in case the local variable is optimiz…

### PR DESCRIPTION
**Pull request by AlexeySotkin**
…ed out

### CODE REVIEW

File: lib/SPIRV/SPIRVToLLVMDbgTran.cpp [hunk 1]
```diff
@@ -893,6 +893,21 @@ SPIRVToLLVMDbgTran::transDebugIntrinsic(const SPIRVExtInst *DebugInst,
   case SPIRVDebug::Declare: {
     using namespace SPIRVDebug::Operand::DebugDeclare;
     auto LocalVar = GetLocalVar(Ops[DebugLocalVarIdx]);
+    if (getDbgInst<SPIRVDebug::DebugInfoNone>(Ops[VariableIdx])) {
+      // If we don't have the variable(e.g. alloca might be promoted by mem2reg)
+      // we should generate the following IR:
+      // call void @llvm.dbg.declare(metadata !4, metadata !14, metadata !5)
+      // !4 = !{}
+      // DIBuilder::insertDeclare doesn't allow to pass nullptr for the Storage
+      // paramter. To work around this limitation we create a dummy temp alloca,
```
Comment (svenvh) [inline lib/SPIRV/SPIRVToLLVMDbgTran.cpp]: paramter -> parameter

File: lib/SPIRV/SPIRVToLLVMDbgTran.cpp [hunk 2]
```diff
@@ -893,6 +893,21 @@ SPIRVToLLVMDbgTran::transDebugIntrinsic(const SPIRVExtInst *DebugInst,
...
+      // use it to create llvm.dbg.declare, and then remove the alloca.
```
Comment (svenvh) [inline lib/SPIRV/SPIRVToLLVMDbgTran.cpp]: My DebugInfo knowledge is limited, but would it not be better to create an `llvm.dbg.value` call instead, as that seems to be what llvm's mem2reg is doing?

File: lib/SPIRV/SPIRVToLLVMDbgTran.cpp (hunk 1 above)
Comment (AlexeySotkin) [reply lib/SPIRV/SPIRVToLLVMDbgTran.cpp]: Fixed

File: lib/SPIRV/SPIRVToLLVMDbgTran.cpp (hunk 2 above)
Comment (AlexeySotkin) [reply lib/SPIRV/SPIRVToLLVMDbgTran.cpp]: mem2reg creates `llvm.dbg.value` for initialized variable. For uninitialized variables mem2reg just removes alloca and leaves `llvm.deb.declare`

**Pull request merged**